Data/HashMap/Base.hs

```diff
@@ -777,89 +866,155 @@ unionWith f = unionWithKey (const f)
 -- | /O(n+m)/ The union of two maps.  If a key occurs in both maps,
 -- the provided function (first argument) will be used to compute the
 -- result.
-unionWithKey :: (Eq k, Hashable k) => (k -> v -> v -> v) -> HashMap k v -> HashMap k v
-          -> HashMap k v
-unionWithKey f = go 0
+unionWithKey
+    :: (Eq k, Hashable k)
+    => (k -> v -> v -> v)
+    -> HashMap k v
+    -> HashMap k v
+    -> HashMap k v
+unionWithKey f (HashMap sz m) hw =
+    let (diff, m') = unionWithKeyInternal f m hw
+    in HashMap (diff + sz) m'
+{-# INLINE unionWithKey #-}
+
+-- | /O(n+m)/ The union of two maps.  If a key occurs in both maps,
+-- the provided function (first argument) will be used to compute the
+-- result.
+-- Returns a tuple where the first component is how many elements were added
+-- to the first hashmap and the second is the union hashmap itself.
+unionWithKeyInternal
+    :: forall k v . (Eq k, Hashable k)
+    => (k -> v -> v -> v)
+    -> HashMapInner k v
+    -> HashMap k v
+    -> (Int, HashMapInner k v)
```

Comment:
Why the pair? Why not a `HashMap`?

Reply:
Hmm, I can forego all the `Internal` functions and do the size tracking directly inside the function.
The `Internal` approach was a good way to start refactoring things, perhaps working directly with the new `HashMap` directly inside the `go` recursive helpers is better.

It's similar to what you mentioned with

`data RunRes e = RunRes !Int !(Array e)`.

I'll do this refactoring.

Reply:
I just mean that `(Int, HashMapInner)` looks like a poor man's `HashMap`. Am I missing something?

Reply:
@treeowl not at all, this is the sort of obvious thing a review would help me find.

Reply:
@treeowl nevermind, there's a reason why I can't use `HashMap` here. The `go` recursive helpers sometimes have to called on a `HashMapInner` (see here https://github.com/rockbmb/unordered-containers/blob/21312f186949665db06940ba03dd0eacd60fd298/Data/HashMap/Base.hs#L671), which means if I expect its return value to be a `HashMap`, then its argument must also be so.

This is because If its argument must be a `HashMap`, then the line I posted won't work because then each level of the tree needs its own `Int` representing the size of the hashmap below, we cannot do that.

However, given your previous comment I understand that/why using regular tuples here is bad. Maybe something like `data Pair = Pair !Int !HashMapInner` to be passed around by `go` would be better?

Reply:
Don't call it `Pair`, but sure, something like that.